File: DataAlgo_forPy/Linked_Lists/linked_list_nth_last.js
// Write a function that takes a head node and an integer value n
// and then returns the nth to last node in the linked list.
// For example, given:

class Node {
  constructor(value) {
    this.nextNode = null
    this.value = value
  }
}

const nth_to_last_node = (k, node) => {
  let marker = node

  for (var i = 0; i < k; i++) {
    marker = node.nextNode
  }

  while (marker.nextNode) {
    node = node.nextNode
    marker = marker.nextNode
  }

  return node.value
}

const a = new Node(1)
const b = new Node(2)
const c = new Node(3)
const d = new Node(4)
const e = new Node(5)

a.nextNode = b
b.nextNode = c
c.nextNode = d
d.nextNode = e

// This would return the node d with a value of 4, because its the 2nd to last node.
const target_node = nth_to_last_node(2, a)
console.log(target_node) // 4

File: DataAlgo_forPy/Linked_Lists/linked_list_nth_last.rb
# Write a function that takes a head node and an integer value n
# and then returns the nth to last node in the linked list.
# For example, given:

class Node
  attr_accessor :nextNode
  attr_reader :value

  def initialize(value)
    @nextNode = nil
    @value = value
  end
end

def nth_to_last_node(k, node)
  marker = node

  (k - 1).times { marker = marker.nextNode }

  while marker.nextNode
    node = node.nextNode
    marker = marker.nextNode
  end

  node.value
end

a = Node.new(1)
b = Node.new(2)
c = Node.new(3)
d = Node.new(4)
e = Node.new(5)

a.nextNode = b
b.nextNode = c
c.nextNode = d
d.nextNode = e

# This would return the node d with a value of 4, because its the 2nd to last node.
target_node = nth_to_last_node(2, a)
p target_node # 4
